fgpyo/collections/__init__.py

```diff
@@ -47,29 +64,60 @@
     >>> next(piter)
     StopIteration
 
-`PeekableIterator`'s constructor supports creation from iterable objects as well as iterators.
+:class:`~fgpyo.collections.PeekableIterator`'s constructor supports creation from iterable objects
+as well as iterators.
 
 Module Contents
 ~~~~~~~~~~~~~~~
 
 The module contains the following public classes:
 
     - :class:`~fgpyo.collections.PeekableIterator` -- Iterator that allows you to peek at the
-        next value before calling next
+      next value without advancing the iterator.
+
+The module contains the following public functions:
+
+    - :func:`~fgpyo.collections.is_sorted` -- for lazily checking if an iterable is sorted.
 
 """
 
+import sys
+from operator import le
 from typing import Any
 from typing import Callable
 from typing import Generic
 from typing import Iterable
 from typing import Iterator
 from typing import List
+from typing import Protocol
+from typing import Tuple
 from typing import TypeVar
 from typing import Union
 
+if sys.version_info[:2] >= (3, 10):
+    from itertools import pairwise
+else:
+    # TODO: remove this branch when Python <3.10 support is dropped
+    def pairwise(iterable: Iterable[Any]) -> Iterator[Tuple[Any, Any]]:
+        """Return successive overlapping pairs taken from the input iterable."""
+        iterator = iter(iterable)
+        head = next(iterator, None)
+        for other in iterator:
+            yield head, other
+            head = other
+
+
+class _SupportsLessThanOrEqual(Protocol):
+    """A structural type for objects that support less-than-or-equal comparison."""
+
+    def __le__(self, other: Any) -> bool: ...
+
+
 IterType = TypeVar("IterType")
 
+_LessThanOrEqualType = TypeVar("_LessThanOrEqualType", bound=_SupportsLessThanOrEqual)
```

Comment:
Is it appropriate for this type be "private" when it is used as a type parameter in a "public" method below?

Reply:
Can do!

I notice fgpyo doesn't use `__init__.py` for public namespace definition which makes defining a public API trickier. For example, the function `pairwise` in this PR is now exposed as our public API instead of just as an import or a helper. I do see Matt is tracking that issue here, though:

- https://github.com/fulcrumgenomics/fgpyo/issues/136

Reply:
Agreed, one way to manage what is public versus private is #136, but another way to solve it is to the way we're doing here.  Moving things out of `__init__.py` into submodules doesn't solve this, since the methods/classes can be public there, but it does help with namespace issues.  Lets move that discussion to #136 since the namespace issue is compelling, and any new method can always be private like we have here for pairwise.